compiler/src/Gren/Format.hs

```diff
@@ -0,0 +1,760 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedLists #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-error=unused-matches #-}
+
+module Gren.Format (toByteStringBuilder) where
+
+import AST.Source qualified as Src
+import AST.Utils.Binop qualified as Binop
+import Control.Monad (join)
+import Data.Bifunctor (second)
+import Data.ByteString.Builder qualified as B
+import Data.List qualified as List
+import Data.List.NonEmpty (NonEmpty ((:|)), nonEmpty)
+import Data.List.NonEmpty qualified as NonEmpty
+import Data.Map.Strict (Map)
+import Data.Map.Strict qualified as Map
+import Data.Maybe (catMaybes)
+import Data.Name (Name)
+import Data.Semigroup (sconcat)
+import Data.Utf8 qualified as Utf8
+import Parse.Primitives qualified as P
+import Reporting.Annotation qualified as A
+import Text.PrettyPrint.Avh4.Block (Block)
+import Text.PrettyPrint.Avh4.Block qualified as Block
+
+toByteStringBuilder :: Src.Module -> B.Builder
+toByteStringBuilder module_ =
+  Block.render (formatModule module_)
+
+--
+-- Data structure extras
+--
+
+repair :: [(a, b)] -> a -> (a, [(b, a)])
+repair [] a' = (a', [])
+repair ((first, b) : rest) a' =
+  (first, repairHelp b rest a')
+  where
+    repairHelp b1 [] a1 = [(b1, a1)]
+    repairHelp b1 ((a1, b2) : rest') a2 =
+      (b1, a1) : repairHelp b2 rest' a2
+
+--
+-- Helper functions
+--
+
+utf8 :: Utf8.Utf8 any -> Block.Line
+utf8 = Block.lineFromBuilder . Utf8.toBuilder
+
+addBlankLines :: Int -> Block -> Block
+addBlankLines n block =
+  Block.stack $
+    NonEmpty.prependList
+      (replicate n Block.blankLine)
+      (NonEmpty.singleton block)
+
+spaceOrStack :: NonEmpty Block -> Block
+spaceOrStack = Block.rowOrStack (Just Block.space)
+
+spaceOrIndent :: NonEmpty Block -> Block
+spaceOrIndent = Block.rowOrIndent (Just Block.space)
+
+spaceOrIndent' :: Bool -> NonEmpty Block -> Block
+spaceOrIndent' forceMultiline = Block.rowOrIndent' forceMultiline (Just Block.space)
+
+{-# INLINE group #-}
```

Comment:
Does this make a big difference, performance wise?

Reply:
Probably not.  My estimation is there's maybe a 60% chance this does improve performance at the cost of increased binary size, since this is a large enough function that ghc might not inline it without the hint.  But also the improvement is certainly going to be tiny and probably not noticable.

Reply:
Would you mind removing the pragmas, then?

Don't mind keeping it if we know for certain that it makes an important difference in performance, but until we know that to be true I'd rather just trust the compiler to do its thing.